File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.java
import java.util.Scanner;

class Main {
    static int N, K, ans;
    static int[] arr;

    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
            N = sc.nextInt();			// 원소 개수
            K = sc.nextInt();			// 부분집합의 합
            arr = new int[N];
            for(int i = 0; i < N; i++) {
                arr[i] = sc.nextInt();	// 수열 입력받기
            }
            ans = 0;
            dfs(0, 0);
        if (K == 0) System.out.println(ans - 1); 
        //S 0일때가 전체 합이 0 일때랑 겹치기 때문에 -1 을 해주어야 합니다.
         else  System.out.println(ans);
        
    }
    static void dfs(int idx, int sum) {
        if(idx == N) {					// 수열을 전부 체크했을 때
            if(sum == K) ans++;		// 부분집합의 합이 K와 같으면
            return;
        } else {
            // 부분집합에 포함시키지 않음 -> 더하지 않음
            dfs(idx + 1, sum);
            // 부분집합에 포함시킴 -> 더함
            dfs(idx + 1, sum+arr[idx]);
        }
    }
}

File: 백준/Silver/1182. 부분수열의 합/부분수열의 합.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(filePath).toString().trim().split(/\s+/);

let idx = 0;
const n = Number(input[idx++]);
const s = Number(input[idx++]);
const arr = [];
let ans=0;

for(let i=0; i<n; i++) {
    arr.push(Number(input[idx++]));
}

dfs(0, 0);

function dfs(index, sum) {
    if(index==n) {
        if(sum===s) ans++;
        return;
    }
    
    // 현재 원소를 포함하는 경우
    dfs(index+1, sum+arr[index]);
    // 현재 원소를 포함하지 않는 경우
    dfs(index+1, sum);
}

// 목표 합 s가 0일 때, 아무것도 선택하지 않은 '공집합' 케이스 제외
if(s===0) {
    console.log(ans-1);
} else {
    console.log(ans);
}